[Skylark] Make range function lazy.

ttsugriy opened the pull request:
range used to use MutableList which would eagerly allocate an array list with
all range elements, which is not efficient for very large ranges or when only
a small number of its elements are used.

This implementation uses a constant amount of RAM and computes a value for
each requested index.

For the following Skylark snippet:
```
def check_content(t):
  if t == []:
    return t
  return False

def modulo(n):
  return n % 797

N = 10000000

[check_content(i) for i in range(N)]
[check_content(i) for i in range(N)]
[modulo(i) for i in range(N)]
[modulo(i) for i in range(N)]
```
the total runtime goes from
```
$ time bazel-bin/src/main/java/com/google/devtools/skylark/Skylark test.bzl
bazel-bin/src/main/java/com/google/devtools/skylark/Skylark test.bzl  93.09s user 1.67s system 316% cpu 29.930 total
```
to
```
$ time bazel-bin/src/main/java/com/google/devtools/skylark/Skylark test.bzl
bazel-bin/src/main/java/com/google/devtools/skylark/Skylark test.bzl  31.45s user 0.86s system 179% cpu 17.974 total
```
which reflects the reduced system time (fewer allocations) and performance.

johnynek commented:
This is exciting!

laurentlb commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +207,178 @@ public int hashCode() {
+      for (int pos : sliceIndices) {
+        builder.add(this.get(pos));
+      }
+      return MutableList.createImmutable(builder.build());
```
Maybe implement it lazily?
https://github.com/google/skylark/blob/master/library.go#L888

laurentlb replied:
A TODO is fine for me.
Be careful with lazy slices, e.g.

```
a = [1, 2, 3]
b = a[1:]
a[2] = 5
b # [2, 3]
```

laurentlb commented on src/test/java/com/google/devtools/build/lib/syntax/MethodLibraryTest.java:
```diff
@@ -486,6 +486,7 @@ public void testRange() throws Exception {
         .testStatement("str(range(5, 0, -1))", "[5, 4, 3, 2, 1]")
         .testStatement("str(range(5, 0, -10))", "[5]")
         .testStatement("str(range(0, -3, -2))", "[0, -2]")
+        .testStatement("str(range(3)[1:2])", "[1]")
```
Please add more tests combining range and slicing
https://github.com/google/skylark/blob/master/testdata/builtins.sky#L94

laurentlb commented:
Thanks a lot! It looks good.

I'll be away a few days, deferring to Jon for the approval

brandjon commented:
Thanks for the PR. I see a couple issues.

The first is that we should try to preserve backward compatibility until we're ready to introduce a behavioral change in a controlled way (i.e. with an --incompatible_* change flag). For range(), the current behavior is that it returns a list, but in the future we want it to return a view. The difference might be observed when the user does any of the following on the result of range():

- takes its str()/repr() or type()
- compares it to other range() results or to normal lists/tuples, with respect to either equality or order
- mutates it
- concatenates it with lists or tuples
- takes a slice of it and then does any of the above

I've added a quick test to check some of these (33f08e7). In any case, ideally they should all be maintained across any semantics-preserving optimization PR like this one.

(We have an old unimplemented internal design doc on the desired view-like behavior of range(). We really should publish that and revisit the design so we all agree on what the end state will be. Created #5264 to track.)

The second issue is the class hierarchy of non-list-non-tuple sequences. The pattern we're trying to evolve toward is that each distinct user-visible Skylark type corresponds to a class with a `@SkylarkModule` annotation. For example, SkylarkList is "sequence", MutableList is "list", Tuple is "tuple". A subclass should not have its own `@SkylarkModule` annotation unless its type is observably different (e.g., a different type() or additional `@SkylarkCallable` API). See the very recent explanation added in cdf5a07.

In this PR, given the backward compatibility constraints that range() objects should (for now) appear to be lists, this means that ideally RangeList and MutableList should inherit from a common interface (ListLike?) that is a SkylarkList, and the "list" `@SkylarkModule` should be pushed up to this interface. range() objects should not be concatable with tuples because lists are not. I think that means we can eliminate TupleLike.

To make progress, I suggest we make sure we have test coverage for range-list-compatibility on the behaviors I listed above (we might already be good, I haven't checked in detail), and then modify this PR to comply.

brandjon commented:
> I suggest we make sure we have test coverage for range-list-compatibility on the behaviors I listed above

In fact, I'll do that now.

brandjon commented:
Done in 585418d.

ttsugriy commented:
Thanks for such a detailed information, @brandjon! I'll rebase and see if any of the tests you've kindly added are broken. I'll also happily remove the `TupleLike` interface :)

ttsugriy commented:
after taking another look at the code, I think that maybe it would be better to just introduce the `--incompatible_use_range_type` flag that would be used to decide which implementation to use and add `@SkylarkModule` with a `range` type to the `RangeList` class, since implementing `append` for `RangeList` is going to be messy. What do you think, @brandjon?

ttsugriy commented:
I updated the change to use `--incompatible_range_type` to control whether the new implementation of range should be used and added corresponding tests. For better compatibility with Python 3 though, I guess I still need to reimplement `getSlice` to return `range` instead of `list`...

laurentlb commented:
Looks good to me, thanks!

ttsugriy commented on src/main/java/com/google/devtools/build/lib/syntax/MethodLibrary.java:
```diff
@@ -673,19 +674,13 @@ public Integer invoke(String value) throws EvalException {
-            }
+            // simplify handling of a case like range(-2) or its equivalent range(-2, 0)
+            // by turning it into range(0, 0)
+            stop = Math.max(start, stop);
```
after playing with Python interpreter I realized that Python actually allows this and displays the range the way it was constructed, so this change is not good. I'm going to remove it.

ttsugriy commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,226 @@ public int hashCode() {
+        }
+      }
+
+      private static int computeSize(int start, int stop, int step) {
```
I've decided to implement it similarly to https://github.com/python/cpython/blob/09bb918a61031377d720f1a0fa1fe53c962791b6/Objects/rangeobject.c#L144-L210 to make it easier to port CPython behavior

ttsugriy commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,226 @@ public int hashCode() {
+        final int length = Math.abs(stop - start);
+        final int absolute_step = Math.abs(step);
+        // round up (length / absolute_step) without using floats
+        return 1 + ((length - 1) / absolute_step);
```
replaced with logic from https://github.com/python/cpython/blob/09bb918a61031377d720f1a0fa1fe53c962791b6/Objects/rangeobject.c#L144-L210

ttsugriy commented:
@brandjon, yeah, I just started working on it. I hope to get working slices in the next day or two if my oncall rotation is not too crazy :)

brandjon commented:
I forgot, when you add lazy slicing, can you please also document the incompatible change flag? It goes in site/docs/skylark/backward-compatibility.md.

ttsugriy commented:
Will do, thanks for the pointers @brandjon! I've also updated PR to include lazy range slices.

ttsugriy commented:
added a section to backwards compatibility doc.

brandjon commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+      RangeListView.Slice slice = RangeListView.Slice.from(size(), start, end, step, loc);
+      int substep = slice.step * this.step;
+      int substart = get(slice.start);
+      int substop = get(slice.stop);
```
It's really confusing which indices are inclusive and exclusive in the new code. A comment or two might help. The simplest interpretation is that `start` and `stop` are consistently inclusive and exclusive, throughout. In that case, `slice.stop` is not necessarily a valid input to `get()`, when the slice's endpoint is the same as the original range's.

ttsugriy replied:
Excellent point! :) I've updated code to use [start, stop) convention everywhere

brandjon commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+      @Override
+      public Integer get(int index) {
+        int value = start + step * index;
+        if ((step > 0 && value > stop) || (step < 0 && value < stop)) {
```
If stop is exclusive, shouldn't it be if `value >= stop` and `value <= stop`?

Then again, why do the computation on values at all? Why not just test if `index < 0 || index >= size`?

ttsugriy replied:
great point! I didn't realize that index that is passed here is already normalized

brandjon commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+       * Python version: https://github.com/python/cpython/blob/09bb918a61031377d720f1a0fa1fe53c962791b6/Objects/rangeobject.c#L144
+       */
+      private static int computeSize(int start, int stop, int step) {
+        int low;
```
In light of the inclusion/exclusion confusion elsewhere, can you add a comment indicating that this is a half-open interval and it doesn't matter which var represents the included endpoint?

brandjon commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+            throw new EvalException(loc, "slice step cannot be zero");
+          }
+
+          int upper;
```
Comment that one endpoint is inclusive and the other exclusive.

brandjon commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+          int lower;
+          if (step < 0) {
+            lower = -1;
+            upper = length + lower;
```
`length - 1` is clearer than `length - lower`.

ttsugriy commented:
thanks for the awesome suggestions, @brandjon! I've addressed all of them and added a couple of test cases. Please take another look.

ttsugriy commented:
@brandjon, could you please take another look?

laurentlb commented:
Sorry for the delay - @brandjon is on leave.
I think the PR is good to go, I'm importing it. If Jon wanted more changes, we can do that later (because the change is behind a flag).

Thanks a lot for your contribution, it's very exciting!

ttsugriy commented:
thank you, @laurentlb! I'll be happy to address any concerns Jon may have once he's back :)

c-parsons commented on src/test/java/com/google/devtools/build/lib/syntax/MethodLibraryTest.java:
```diff
@@ -486,6 +486,7 @@ public void testRange() throws Exception {
         .testStatement("str(range(5, 0, -1))", "[5, 4, 3, 2, 1]")
```
Add a test for mult, which is also now disabled?

ttsugriy replied:
I thought I had one... I guess I was wrong... :)

c-parsons commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,328 @@ public int hashCode() {
+              + "range(10)[::2]  # range(0, 10, 2)\n"
+              + "range(10)[3:0:-1]  # range(3, 0, -1)</pre>"
+              + "Ranges are immutable, as in Python 3.")
+  public static final class RangeList extends SkylarkList<Integer> {
```
I have a preference for these new classes be in separate files. Please :)

ttsugriy replied:
yeah, I just wasn't sure what's the convention, since the rest of SkylarkList implementations are inlined :)

c-parsons commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,328 @@ public int hashCode() {
+    private static class RangeListView extends AbstractList<Integer> {
+
+      /** Iterator for increasing sequences. */
+      private static class IncreasingIterator extends UnmodifiableIterator<Integer> {
```
In principle I thought it was fine that we were separating IncreasingIterator and DecreasingIterator for clarity purposes, but then it turns out that these classes are identical except for hasNext().

Perhaps we should combine them, and hasNext() can be   return (step > 0) ? : cursor < stop : cursor > stop;

ttsugriy replied:
that's how I started, but then started prematurely optimizing to avoid extra branches :)

c-parsons commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,328 @@ public int hashCode() {
+      }
+    }
+
+    public MutableList<Integer> toMutableList(Environment env) {
```
javadoc? (at least to describe the ramifications of doing this)javadoc

c-parsons commented on src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java:
```diff
@@ -201,6 +206,328 @@ public int hashCode() {
+      return MutableList.copyOf(env, contents);
+    }
+
+    public static RangeList of(int start, int stop, int step) {
```
javadoc

ttsugriy commented:
addressed all concerns, @c-parsons, please take another look.

c-parsons commented on src/main/java/com/google/devtools/build/lib/syntax/RangeList.java:
```diff
@@ -0,0 +1,312 @@
+  /** Provides access to range elements based on their index. */
+  private static class RangeListView extends AbstractList<Integer> {
+
+    /** Iterator for increasing sequences. */
```
Fix javadoc

ttsugriy replied:
doh

c-parsons commented on src/main/java/com/google/devtools/build/lib/syntax/RangeList.java:
```diff
@@ -0,0 +1,312 @@
+  }
+
+  /**
+   * @return A materialized version of the range that can be used as a <pre>list</pre> type.<p>
```
Nit: Throughout bazel we put @return at the end of javadoc blocks

ttsugriy replied:
updated

ttsugriy commented:
addressed all of the comments above, please take another look, @c-parsons :)

ttsugriy commented:
Thank you for the review, @c-parsons!
